File: php/src/Entity/Article.php
<?php

declare(strict_types=1);

namespace App\Entity;

use App\Repository\ArticleRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Bridge\Doctrine\Validator\Constraints\UniqueEntity;

#[ORM\Entity(repositoryClass: ArticleRepository::class)]
#[ORM\Table(name: 'articles')]
#[ORM\UniqueConstraint(name: 'uniq_articles_slug', columns: ['slug'])]
#[UniqueEntity(fields: ['slug'])]
#[ORM\HasLifecycleCallbacks]
class Article
{
    public const STATUS_DRAFT = 'draft';
    public const STATUS_REVIEW = 'review';
    public const STATUS_SCHEDULED = 'scheduled';
    public const STATUS_PUBLISHED = 'published';
    public const STATUS_ARCHIVED = 'archived';

    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(length: 255)]
    private string $title = '';

    #[ORM\Column(length: 255)]
    private string $slug = '';

    #[ORM\Column(type: 'text', nullable: true)]
    private ?string $description = null;

    #[ORM\Column(type: 'text')]
    private string $content = '';

    #[ORM\Column(length: 20)]
    private string $status = self::STATUS_DRAFT;

    #[ORM\ManyToOne]
    #[ORM\JoinColumn(nullable: false)]
    private ?User $author = null;

    #[ORM\Column(nullable: true)]
    private ?\DateTimeImmutable $publishedAt = null;

    #[ORM\Column]
    private \DateTimeImmutable $createdAt;

    #[ORM\Column]
    private \DateTimeImmutable $updatedAt;

    #[ORM\ManyToOne]
    private ?Media $featuredImage = null;

    #[ORM\Column(options: ['default' => false])]
    private bool $isBreaking = false;

    #[ORM\Column(options: ['default' => 0])]
    private int $viewCount = 0;

    /**
     * @var Collection<int, Category>
     */
    #[ORM\ManyToMany(targetEntity: Category::class, inversedBy: 'articles')]
    #[ORM\JoinTable(name: 'article_categories')]
    private Collection $categories;

    /**
     * @var Collection<int, Tag>
     */
    #[ORM\ManyToMany(targetEntity: Tag::class, inversedBy: 'articles')]
    #[ORM\JoinTable(name: 'article_tags')]
    private Collection $tags;

    /**
     * @var Collection<int, ArticleRevision>
     */
    #[ORM\OneToMany(targetEntity: ArticleRevision::class, mappedBy: 'article')]
    private Collection $revisions;

    public function __construct()
    {
        $this->categories = new ArrayCollection();
        $this->tags = new ArrayCollection();
        $this->revisions = new ArrayCollection();
    }

    #[ORM\PrePersist]
    public function prePersist(): void
    {
        $now = new \DateTimeImmutable();
        $this->createdAt = $now;
        $this->updatedAt = $now;
    }

    #[ORM\PreUpdate]
    public function preUpdate(): void
    {
        $this->updatedAt = new \DateTimeImmutable();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getTitle(): string
    {
        return $this->title;
    }

    public function setTitle(string $title): self
    {
        $this->title = $title;

        return $this;
    }

    public function getSlug(): string
    {
        return $this->slug;
    }

    public function setSlug(string $slug): self
    {
        $this->slug = $slug;

        return $this;
    }

    public function getDescription(): ?string
    {
        return $this->description;
    }

    public function setDescription(?string $description): self
    {
        $this->description = $description;

        return $this;
    }

    public function getContent(): string
    {
        return $this->content;
    }

    public function setContent(string $content): self
    {
        $this->content = $content;

        return $this;
    }

    public function getStatus(): string
    {
        return $this->status;
    }

    public function setStatus(string $status): self
    {
        $this->status = $status;

        return $this;
    }

    public function getAuthor(): ?User
    {
        return $this->author;
    }

    public function setAuthor(?User $author): self
    {
        $this->author = $author;

        return $this;
    }

    public function getPublishedAt(): ?\DateTimeImmutable
    {
        return $this->publishedAt;
    }

    public function setPublishedAt(?\DateTimeImmutable $publishedAt): self
    {
        $this->publishedAt = $publishedAt;

        return $this;
    }

    public function getCreatedAt(): \DateTimeImmutable
    {
        return $this->createdAt;
    }

    public function getUpdatedAt(): \DateTimeImmutable
    {
        return $this->updatedAt;
    }

    public function getFeaturedImage(): ?Media
    {
        return $this->featuredImage;
    }

    public function setFeaturedImage(?Media $featuredImage): self
    {
        $this->featuredImage = $featuredImage;

        return $this;
    }

    public function isBreaking(): bool
    {
        return $this->isBreaking;
    }

    public function setIsBreaking(bool $isBreaking): self
    {
        $this->isBreaking = $isBreaking;

        return $this;
    }

    public function getViewCount(): int
    {
        return $this->viewCount;
    }

    public function setViewCount(int $viewCount): self
    {
        $this->viewCount = $viewCount;

        return $this;
    }

    /**
     * @return Collection<int, Category>
     */
    public function getCategories(): Collection
    {
        return $this->categories;
    }

    public function addCategory(Category $category): self
    {
        if (!$this->categories->contains($category)) {
            $this->categories->add($category);
        }

        return $this;
    }

    public function removeCategory(Category $category): self
    {
        $this->categories->removeElement($category);

        return $this;
    }

    /**
     * @return Collection<int, Tag>
     */
    public function getTags(): Collection
    {
        return $this->tags;
    }

    public function addTag(Tag $tag): self
    {
        if (!$this->tags->contains($tag)) {
            $this->tags->add($tag);
        }

        return $this;
    }

    public function removeTag(Tag $tag): self
    {
        $this->tags->removeElement($tag);

        return $this;
    }

    /**
     * @return Collection<int, ArticleRevision>
     */
    public function getRevisions(): Collection
    {
        return $this->revisions;
    }

    /**
     * @return list<string>
     */
    public static function availableStatuses(): array
    {
        return [
            self::STATUS_DRAFT,
            self::STATUS_REVIEW,
            self::STATUS_SCHEDULED,
            self::STATUS_PUBLISHED,
            self::STATUS_ARCHIVED,
        ];
    }

    public function __toString(): string
    {
        return $this->title;
    }
}

File: elixir/news_umbrella/apps/core/lib/core/news/article.ex
defmodule Core.News.Article do
  use Ecto.Schema
  import Ecto.Changeset
  alias Core.News.ChangesetHelpers

  @statuses ~w(draft review scheduled published archived)

  schema "articles" do
    field(:title, :string)
    field(:slug, :string)
    field(:description, :string)
    field(:content, :string)
    field(:status, :string, default: "draft")
    field(:published_at, :utc_datetime)
    field(:is_breaking, :boolean, default: false)
    field(:view_count, :integer, default: 0)

    field(:author, :string)
    belongs_to(:featured_image, Core.News.Media)

    many_to_many(:categories, Core.News.Category,
      join_through: "article_categories",
      on_replace: :delete
    )

    many_to_many(:tags, Core.News.Tag,
      join_through: "article_tags",
      on_replace: :delete
    )

    has_many(:revisions, Core.News.ArticleRevision)

    timestamps(type: :utc_datetime)
  end

  def changeset(article, attrs) do
    article
    |> cast(attrs, [
      :title,
      :slug,
      :description,
      :content,
      :status,
      :published_at,
      :is_breaking,
      :view_count,
      :author,
      :featured_image_id
    ])
    |> ChangesetHelpers.trim_string_fields([
      :title,
      :slug,
      :description,
      :content,
      :status,
      :author
    ])
    |> validate_required([:title, :slug, :content, :status, :author])
    |> validate_length(:title, min: 5, max: 180)
    |> ChangesetHelpers.validate_slug(:slug, max: 200)
    |> validate_length(:description, max: 1_000)
    |> validate_length(:content, min: 20, max: 120_000)
    |> validate_length(:author, min: 3, max: 120)
    |> validate_inclusion(:status, @statuses)
    |> validate_number(:view_count,
      greater_than_or_equal_to: 0,
      less_than_or_equal_to: 10_000_000
    )
    |> validate_published_at_when_published()
    |> unique_constraint(:slug)
    |> foreign_key_constraint(:featured_image_id)
  end

  defp validate_published_at_when_published(changeset) do
    status = get_field(changeset, :status)
    published_at = get_field(changeset, :published_at)

    if status == "published" and is_nil(published_at) do
      add_error(changeset, :published_at, "is required when status is published")
    else
      changeset
    end
  end
end
